apps/io-func-wallet-solution-backend/src/infra/attestation-service/android/assertion.ts

```diff
@@ -0,0 +1,206 @@
+import { JWK, KeyLike, exportSPKI, importJWK } from "jose";
+
+const ALLOWED_WINDOW_MILLIS = 1000 * 60; // 60 seconds
+
+export const GoogleAppCredentials = t.type({
+  type: NonEmptyString,
+  project_id: NonEmptyString,
+  private_key_id: NonEmptyString,
+  private_key: NonEmptyString,
+  client_email: EmailString,
+  client_id: NonEmptyString,
+  auth_uri: NonEmptyString,
+  token_uri: NonEmptyString,
+  auth_provider_x509_cert_url: NonEmptyString,
+  client_x509_cert_url: NonEmptyString,
+  universe_domain: NonEmptyString,
+});
+
+export type GoogleAppCredentials = t.TypeOf<typeof GoogleAppCredentials>;
+
+export type VerifyAssertionParams = {
+  integrityAssertion: NonEmptyString;
+  hardwareSignature: string;
+  clientData: string;
+  hardwareKey: JWK;
+  bundleIdentifier: string;
+  googleAppCredentials: GoogleAppCredentials;
+  androidPlayIntegrityUrl: string;
+};
+
+export const playintegrity = google.playintegrity("v1");
+
+export const verifyAssertion = async (params: VerifyAssertionParams) => {
+  const {
+    integrityAssertion,
+    hardwareSignature,
+    clientData,
+    hardwareKey,
+    bundleIdentifier,
+    googleAppCredentials,
+    androidPlayIntegrityUrl,
+  } = params;
+
+  // First check whether the clientData has been signed correctly with the hardware key
+  const signatureValidated = validateAssertionSignature(
+    hardwareKey,
+    clientData,
+    hardwareSignature
+  );
+
+  if (!signatureValidated) {
+    throw new Error("Invalid hardware signature");
+  }
+
+  // Then verify the integrity token
+  const jwtClient = new google.auth.JWT(
+    googleAppCredentials.client_email,
+    undefined,
+    googleAppCredentials.private_key,
+    [androidPlayIntegrityUrl]
+  );
+
+  google.options({ auth: jwtClient });
+  const result = await playintegrity.v1.decodeIntegrityToken({
+    packageName: bundleIdentifier,
+    requestBody: {
+      integrityToken: integrityAssertion,
+    },
+  });
+
+  const tokenPayloadExternal = result.data.tokenPayloadExternal;
+
+  if (!tokenPayloadExternal) {
+    throw new Error("Invalid token payload from Play Integrity API response");
+  }
+
+  const responseValidated = validateIntegrityResponse(
+    tokenPayloadExternal,
+    bundleIdentifier,
+    clientData
+  );
+
+  if (!responseValidated) {
+    throw new Error("Integrity Response did not pass validation");
+  }
+
+  return true;
+};
+
+export const validateAssertionSignature = async (
+  hardwareKey: JWK,
+  clientData: string,
+  hardwareSignature: string
+) => {
+  const joseHardwareKey = (await importJWK(hardwareKey)) as KeyLike;
+  const publicHardwareKeyPem = await exportSPKI(joseHardwareKey);
+
+  const clientDataHash = createHash("sha256").update(clientData).digest();
+
+  const verifier = createVerify("SHA256");
+  verifier.update(clientDataHash);
+  return verifier.verify(publicHardwareKeyPem, hardwareSignature, "base64");
+};
+
+export const validateIntegrityResponse = (
```

Comment:
We are not checking the `certificateSha256Digest`, is this intended? What about the `versionCodex`? We could use it enforce only some version of the app to obtain an attestation without forcing every user of the app to update to the latest version. Does that makes sense to you?

Reply:
I skipped the versionCode check precisely to avoid that some versions of the app could skip the checks, in this way they are all checked and I think it's the best thing to do. Why do you think we need to filter some version from the integrity check? Consider that in any case here the request occurs only when obtaining the Wallet Attestation and the app has already called the endpoints for the integrity check.

For the Sha256Digest certificate I have skipped it for now as we have added the verification of the signature via the attested key for which the certificate has already been verified. But this is a check we could add

Reply:
I was thinking about a way to force the users in the experimentation to have a particular version without forcing every user of the app to do so via our blocking banner in app.  However a more thoughtful thinking made me realise that we couldn't apply this logic to iOS for example, due to the differences in the integrity check. Nevermind!

Regarding the `certificateSha256Digest`, it should be the base64-encoded URL-safe of our app certificate used for publishing it on GPS. This a different check from the one you mentioned, isn't it?

Thank you

Reply:
ok then no, the check must be done. Do we have access to the certificate somehow?

Reply:
At the moment I would say no. That would involve the application being available on GPS. Can we verify its presence within our environment and perform the check accordingly? If it's absent, we can bypass this step.

Reply:
Fixed in https://github.com/pagopa/io-wallet/pull/43/commits/ca0ed13e299270629e2fcb5fe9f054dce8a888ed